File: frontend-v2/src/lib/api/digitalHuman.ts
import { api } from "./client";

export interface DigitalHumanAvatar {
  id: number;
  name: string;
  avatar_type: string;
  thumbnail_url: string | null;
  video_url: string | null;
  gender: string | null;
  is_active: boolean;
}

export interface DigitalHumanVoice {
  id: number;
  name: string;
  voice_type: string;
  sample_url: string | null;
  gender: string | null;
  is_active: boolean;
}

export interface DigitalHumanVideo {
  id: number;
  title: string;
  video_url: string | null;
  audio_url: string | null;
  duration: number | null;
  status: string;
  error_message: string | null;
  created_at: string;
}

export interface GenerateVideoPayload {
  project_id: number;
  script_id: number;
  voice_id: number;
  avatar_id: number;
  with_subtitle?: boolean;
  with_bgm?: boolean;
  resolution?: string;
}

export async function listAvatars(): Promise<DigitalHumanAvatar[]> {
  return api.get<DigitalHumanAvatar[]>("/api/digital-human/avatars");
}

export async function listVoices(): Promise<DigitalHumanVoice[]> {
  return api.get<DigitalHumanVoice[]>("/api/digital-human/voices");
}

export async function cloneVoice(formData: FormData): Promise<DigitalHumanVoice> {
  const response = await fetch(`${process.env.NEXT_PUBLIC_API_BASE_URL || ""}/api/digital-human/voices/clone`, {
    method: "POST",
    body: formData,
    credentials: "include",
  });
  if (!response.ok) {
    const text = await response.text();
    throw new Error(text || "声音克隆失败");
  }
  const json = await response.json();
  if (!json.success) {
    throw new Error(json.message || "声音克隆失败");
  }
  return json.data;
}

export async function listVideos(projectId?: number): Promise<DigitalHumanVideo[]> {
  const params = projectId ? `?project_id=${projectId}` : "";
  return api.get<DigitalHumanVideo[]>(`/api/digital-human/videos${params}`);
}

export async function generateVideo(payload: GenerateVideoPayload): Promise<{ task_id: number; video_id: number; status: string; message: string }> {
  return api.post("/api/digital-human/videos/generate", payload);
}

File: backend/app/schemas/digital_human.py
from pydantic import BaseModel, Field
from datetime import datetime


class DigitalHumanAvatarBase(BaseModel):
    name: str
    avatar_type: str = "preset"
    thumbnail_url: str | None = None
    video_url: str | None = None
    gender: str | None = None
    config_json: dict = Field(default_factory=dict)
    is_active: bool = True


class DigitalHumanAvatarCreate(DigitalHumanAvatarBase):
    pass


class DigitalHumanAvatarRead(DigitalHumanAvatarBase):
    id: int
    created_at: datetime

    class Config:
        from_attributes = True


class DigitalHumanVoiceBase(BaseModel):
    name: str
    voice_type: str = "preset"
    sample_url: str | None = None
    gender: str | None = None
    config_json: dict = Field(default_factory=dict)
    is_active: bool = True


class DigitalHumanVoiceCreate(DigitalHumanVoiceBase):
    pass


class DigitalHumanVoiceRead(DigitalHumanVoiceBase):
    id: int
    user_id: int | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class DigitalHumanVideoBase(BaseModel):
    project_id: int | None = None
    script_id: int | None = None
    voice_id: int | None = None
    avatar_id: int | None = None
    title: str
    video_url: str | None = None
    audio_url: str | None = None
    duration: float | None = None
    status: str = "pending"
    config_json: dict = Field(default_factory=dict)
    error_message: str | None = None
    credit_cost: int | None = None


class DigitalHumanVideoCreate(BaseModel):
    project_id: int
    script_id: int
    voice_id: int
    avatar_id: int
    with_subtitle: bool = True
    with_bgm: bool = False
    resolution: str = "1080p"


class DigitalHumanVideoRead(DigitalHumanVideoBase):
    id: int
    user_id: int | None = None
    task_id: int | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class DigitalHumanVideoGenerateResponse(BaseModel):
    task_id: int
    video_id: int
    status: str
    message: str
